File: quant1x/datasets/xdxr.cpp
#include <quant1x/level1/client.h>
#include <quant1x/datasets/xdxr.h>
#include <quant1x/config/base.h>
#include <quant1x/config/cache.h>
#include <quant1x/std/filepath.h>

namespace datasets {

    std::vector<level1::XdxrInfo> load_xdxr(const std::string& code) {
        std::vector<level1::XdxrInfo> result;
        try {
            auto cache_filename = config::get_xdxr_filename(code);
            io::CSVReader<14> in(cache_filename);
            in.read_header(io::ignore_extra_column, "date", "category", "name", "fen_hong", "pei_gu_jia", "song_zhuan_gu",
                           "pei_gu", "suo_gu", "qian_liu_tong", "hou_liu_tong", "qian_zong_gu_ben", "hou_zong_gu_ben", "fen_shu",
                           "xing_quan_jia");
            level1::XdxrInfo row = {};
            while (in.read_row(row.Date, row.Category, row.Name, row.FenHong, row.PeiGuJia, row.SongZhuanGu, row.PeiGu, row.SuoGu,
                               row.QianLiuTong, row.HouLiuTong, row.QianZongGuBen, row.HouZongGuBen, row.FenShu, row.XingQuanJia)) {
                result.emplace_back(row);

            }
        } catch(...) {
            //
        }
        return result;
    }

    void save_xdxr(const std::string &code, const exchange::timestamp &date, const std::vector<level1::XdxrInfo>& values) {
        std::string securityCode = exchange::CorrectSecurityCode(code);
        exchange::timestamp factor_date = date;
        (void)factor_date;

        auto ofn = config::get_xdxr_filename(securityCode);
        filepath::check_filepath(ofn, true);
        io::CSVWriter writer(ofn);
        writer.write_row("date","category","name","fen_hong","pei_gu_jia","song_zhuan_gu","pei_gu","suo_gu","qian_liu_tong","hou_liu_tong","qian_zong_gu_ben","hou_zong_gu_ben","fen_shu","xing_quan_jia");
        for (const auto &v: values) {
            writer.write_row(v.Date, v.Category, v.Name, v.FenHong,v.PeiGuJia,v.SongZhuanGu,v.PeiGu,v.SuoGu,v.QianLiuTong,v.HouLiuTong,v.QianZongGuBen,v.HouZongGuBen,v.FenShu,v.XingQuanJia);
        }
    }

    cache::Kind DataXdxr::Kind() const  {
        return BaseXdxr;
    }

    std::string DataXdxr::Owner()  {
        return cache::DefaultDataProvider;
    }

    std::string DataXdxr::Key() const  {
        return "xdxr";
    }

    std::string DataXdxr::Name() const  {
        return "除权除息";
    }

    std::string DataXdxr::Usage() const  {
        return std::string();
    }

    void DataXdxr::Print(const std::string &code, const std::vector<exchange::timestamp> &dates)  {
        (void)code;
        (void)dates;
    }

    void DataXdxr::Update(const std::string &code, const exchange::timestamp &date)  {
//        if(date != exchange::last_trading_day()) {
//            return;
//        }
        try {
            auto conn = level1::get_std_conn();
            level1::XdxrInfoRequest request(code);
            level1::XdxrInfoResponse response;
            auto err = level1::process(conn->socket(), request, response);
            if (err) {
                throw std::runtime_error(fmt::format("Process error: {}", err.message()));
            }
            if (response.Count) {
                save_xdxr(code, date, response.List);
            }
        } catch (const std::exception &e) {  // 其他标准异常
            spdlog::error("[dataset::xdxr] - 标准异常: {} (type: {})", e.what(), typeid(e).name());
            // 对于system_error可以记录更多信息
            if (auto se = dynamic_cast<const std::system_error *>(&e)) {
                spdlog::error("[dataset::xdxr] Error code: {}, category: {}", se->code().value(), se->code().category().name());
            }
        } catch (...) {
            spdlog::error("[dataset::xdxr] 获取除权除息异常");
        }
    }

}

File: quant1x/datasets/xdxr.py
# -*- coding: UTF-8 -*-
import logging
import os
import csv
from typing import List

from quant1x.level1 import client, protocol, xdxr_info
from quant1x.exchange import code as exchange_code
from quant1x.exchange import Timestamp
from quant1x.config import config

log = logging.getLogger(__name__)

def load_xdxr(code: str) -> List[xdxr_info.XdxrInfo]:
    result = []
    try:
        filename = config.get_xdxr_filename(code)
        if os.path.exists(filename):
            with open(filename, 'r', encoding='utf-8') as f:
                reader = csv.DictReader(f)
                for row in reader:
                    info = xdxr_info.XdxrInfo()
                    info.Date = row['Date']
                    info.Category = int(row['Category'])
                    info.Name = row['Name']
                    info.FenHong = float(row['FenHong'])
                    info.PeiGuJia = float(row['PeiGuJia'])
                    info.SongZhuanGu = float(row['SongZhuanGu'])
                    info.PeiGu = float(row['PeiGu'])
                    info.SuoGu = float(row['SuoGu'])
                    info.QianLiuTong = float(row['QianLiuTong'])
                    info.HouLiuTong = float(row['HouLiuTong'])
                    info.QianZongGuBen = float(row['QianZongGuBen'])
                    info.HouZongGuBen = float(row['HouZongGuBen'])
                    info.FenShu = float(row['FenShu'])
                    info.XingQuanJia = float(row['XingQuanJia'])
                    result.append(info)
    except Exception as e:
        log.error(f"[dataset::xdxr] Load failed: {e}")
    return result

def save_xdxr(code: str, date: Timestamp, values: List[xdxr_info.XdxrInfo]):
    security_code = exchange_code.correct_security_code(code)
    # date is ignored as in C++ implementation
    
    filename = config.get_xdxr_filename(security_code)
    try:
        os.makedirs(os.path.dirname(filename), exist_ok=True)
        with open(filename, 'w', newline='', encoding='utf-8') as f:
            writer = csv.writer(f)
            writer.writerow(["Date", "Category", "Name", "FenHong", "PeiGuJia", "SongZhuanGu",
                             "PeiGu", "SuoGu", "QianLiuTong", "HouLiuTong", "QianZongGuBen", 
                             "HouZongGuBen", "FenShu", "XingQuanJia"])
            for v in values:
                writer.writerow([v.Date, v.Category, v.Name, v.FenHong, v.PeiGuJia, v.SongZhuanGu,
                                 v.PeiGu, v.SuoGu, v.QianLiuTong, v.HouLiuTong, v.QianZongGuBen,
                                 v.HouZongGuBen, v.FenShu, v.XingQuanJia])
    except Exception as e:
        log.error(f"[dataset::xdxr] Save failed: {e}")

class DataXdxr:
    def kind(self):
        return "BaseXdxr"
        
    def owner(self):
        return "default"
        
    def key(self):
        return "xdxr"
        
    def name(self):
        return "除权除息"
        
    def usage(self):
        return ""
        
    def print(self, code: str, dates: List[Timestamp]):
        pass
        
    def update(self, code: str, date: Timestamp):
        try:
            req = xdxr_info.XdxrInfoRequest(code)
            
            conn = client.client()
            sock = conn.socket
            resp = xdxr_info.XdxrInfoResponse()
            protocol.process(sock, req, resp)
            
            if resp.count > 0:
                save_xdxr(code, date, resp.list)
                
        except Exception as e:
            log.error(f"[dataset::xdxr] 获取除权除息异常: {e}")

